packet/src/pacer.rs

```diff
@@ -0,0 +1,835 @@
+use std::collections::HashMap;
+use std::fmt;
+use std::time::{Duration, Instant};
+
+use rtp::{Bitrate, DataSize};
+
+const MAX_BITRATE: Bitrate = Bitrate::gbps(10);
+const MAX_MEDIA_DEBT_IN_TIME: Duration = Duration::from_millis(500);
+
+/// A packet Pacer.
+///
+/// The pacer is responsible for ensuring correct pacing of packets onto the network at a given
+/// bitrate.
+pub trait Pacer {
+    /// Set the target bitrate to send at. This can be exceeded if required.
+    fn set_target_bitrate(&mut self, target_bitrate: Bitrate);
+
+    /// Poll for a timeout.
+    fn poll_timeout(&self) -> Option<Instant>;
+
+    /// Handle time moving forward, should be called periodically as indicated by `poll_timeout`.
+    fn handle_time(&mut self, now: Instant, queue_states: &[QueueState]);
+
+    /// Determines what action to take, if any.
+    ///
+    ///
+    /// ## Return values
+    /// * [`PollOutcome::PollQueue`] indicates that the queue for which the returned the queue
+    /// state is returned should be polled.
+    /// * [`PollOutcome::Nothing`] indicates no action is required at this time.
+    fn poll_action<'q>(&mut self, now: Instant, queue_states: &'q [QueueState]) -> PollOutcome<'q>;
+
+    /// Register a packet being sent let outbound.
+    ///
+    /// **MUST** be called each time [`Pacer::poll_action`] produces [`PollOutcome::PollQueue`]
+    /// after the packet is sent.
+    fn register_send(&mut self, now: Instant, packet_size: DataSize, from: QueueId);
+}
+
+/// A unique identifier for a given packet queue. Should be immutable.
+#[derive(Debug, Default, PartialEq, Eq, Hash, Clone, Copy)]
+pub struct QueueId(u64);
+
+/// The state of a single upstream queue.
+/// The pacer manages packets across several upstream queues.
+#[derive(Debug)]
+pub struct QueueState {
+    /// The unique identifier for this queue within the session.
+    pub queue_id: QueueId,
+    /// The total size of queued payloads within the queue.
+    pub size: DataSize,
+    /// The total number of packets in the queue.
+    pub packet_count: u32,
```

Comment:
We need to divide it with `Duration` and `Duration / usize` isn't implemented. Rather than using `usize` here and truncating I figured it'd be better to communicate this limitation here.

Reply:
ok. let's do a code note.

Reply:
Sure thing